File: vscode/resources/native_host.js
#!/usr/bin/env node

/**
 * Native Messaging Host for PKM Assistant
 * 
 * This script acts as a bridge between the browser extension and VS Code extension.
 * It uses the Chrome Native Messaging protocol (stdin/stdout with length-prefixed JSON).
 * 
 * Protocol:
 * - Input: 4-byte length header (little-endian) + JSON message
 * - Output: 4-byte length header (little-endian) + JSON response
 */

const fs = require('fs');
const http = require('http');
const path = require('path');
const os = require('os');

// Configuration
const PORT_FILE = path.join(os.homedir(), '.pkm-assistant', 'port.json');
const MAX_MESSAGE_SIZE = 1024 * 1024 * 10; // 10MB max message size

/**
 * Reads a message from stdin using Chrome Native Messaging protocol
 * @returns {Promise<Object>} The parsed JSON message
 */
function readMessage() {
  return new Promise((resolve, reject) => {
    // Read 4-byte length header
    const chunks = [];
    let headerRead = false;
    let messageLength = 0;
    let bytesRead = 0;

    process.stdin.on('readable', function onReadable() {
      let chunk;
      
      // Read header if not yet read
      if (!headerRead) {
        chunk = process.stdin.read(4);
        if (!chunk || chunk.length < 4) {
          return; // Wait for more data
        }
        
        messageLength = chunk.readUInt32LE(0);
        
        if (messageLength > MAX_MESSAGE_SIZE) {
          reject(new Error(`Message too large: ${messageLength} bytes`));
          return;
        }
        
        headerRead = true;
      }
      
      // Read message body
      while ((chunk = process.stdin.read()) !== null) {
        chunks.push(chunk);
        bytesRead += chunk.length;
        
        if (bytesRead >= messageLength) {
          // Remove this listener to prevent further reads
          process.stdin.removeListener('readable', onReadable);
          
          // Concatenate chunks and parse JSON
          const buffer = Buffer.concat(chunks);
          const messageBytes = buffer.slice(0, messageLength);
          
          try {
            const message = JSON.parse(messageBytes.toString('utf8'));
            resolve(message);
          } catch (error) {
            reject(new Error(`Failed to parse message: ${error.message}`));
          }
          break;
        }
      }
    });
  });
}

/**
 * Writes a message to stdout using Chrome Native Messaging protocol
 * @param {Object} message - The message object to send
 */
function writeMessage(message) {
  const messageJson = JSON.stringify(message);
  const messageBuffer = Buffer.from(messageJson, 'utf8');
  
  // Create 4-byte length header
  const headerBuffer = Buffer.allocUnsafe(4);
  headerBuffer.writeUInt32LE(messageBuffer.length, 0);
  
  // Write header and message
  process.stdout.write(headerBuffer);
  process.stdout.write(messageBuffer);
}

/**
 * Reads the VS Code server port from the port file
 * @returns {Promise<number>} The port number
 */
async function readServerPort() {
  try {
    const portData = await fs.promises.readFile(PORT_FILE, 'utf8');
    const portInfo = JSON.parse(portData);
    return portInfo.port;
  } catch (error) {
    throw new Error(`Failed to read port file: ${error.message}`);
  }
}

/**
 * Forwards a message to the VS Code extension server
 * @param {Object} message - The message to forward
 * @param {number} port - The server port
 * @returns {Promise<Object>} The server response
 */
function forwardToVSCode(message, port) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(message);
    
    const options = {
      hostname: 'localhost',
      port: port,
      path: '/visit',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data)
      }
    };
    
    const req = http.request(options, (res) => {
      let responseData = '';
      
      res.on('data', (chunk) => {
        responseData += chunk;
      });
      
      res.on('end', () => {
        try {
          const response = JSON.parse(responseData);
          resolve(response);
        } catch (error) {
          // If response is not JSON, return as-is
          resolve({ status: 'success', data: responseData });
        }
      });
    });
    
    req.on('error', (error) => {
      reject(error);
    });
    
    req.write(data);
    req.end();
  });
}

/**
 * Main message processing loop
 */
async function main() {
  // Log startup for debugging (to stderr to not interfere with stdout protocol)
  console.error('[Native Host] Starting PKM Assistant native messaging host');
  
  // Handle stdin close (browser closed connection)
  process.stdin.on('end', () => {
    console.error('[Native Host] Browser disconnected, exiting');
    process.exit(0);
  });
  
  // Main message loop
  while (true) {
    try {
      // Read message from browser
      const message = await readMessage();
      console.error('[Native Host] Received message:', message.type || 'unknown');
      
      // Handle different message types
      if (message.type === 'ping') {
        // Health check
        writeMessage({ type: 'pong', status: 'healthy' });
      } else if (message.type === 'visit' || !message.type) {
        // Forward webpage visit to VS Code
        const port = await readServerPort();
        const response = await forwardToVSCode(message, port);
        writeMessage({ 
          type: 'response', 
          status: 'success',
          ...response 
        });
      } else {
        // Unknown message type
        writeMessage({ 
          type: 'error', 
          error: `Unknown message type: ${message.type}` 
        });
      }
    } catch (error) {
      console.error('[Native Host] Error:', error.message);
      
      // Send error response to browser
      writeMessage({ 
        type: 'error', 
        status: 'error',
        error: error.message 
      });
      
      // For critical errors, exit
      if (error.message.includes('Failed to read port file')) {
        console.error('[Native Host] VS Code server not running, exiting');
        process.exit(1);
      }
    }
  }
}

// Start the native host
main().catch((error) => {
  console.error('[Native Host] Fatal error:', error);
  process.exit(1);
});

File: native-host/native_host.py
#!/usr/bin/env python3
"""
Bergamot Native Messaging Host
Bridges communication between browser extension and VS Code extension
"""

import sys
import json
import os
import struct
import requests
import logging
from pathlib import Path

# Configure logging
log_dir = Path.home() / '.bergamot'
log_dir.mkdir(exist_ok=True)
logging.basicConfig(
    filename=str(log_dir / 'native-host.log'),
    level=logging.DEBUG,
    format='%(asctime)s - %(levelname)s - %(message)s'
)

def read_message():
    """Read a message from stdin using Chrome native messaging protocol"""
    try:
        # Read the message length (first 4 bytes)
        raw_length = sys.stdin.buffer.read(4)
        if not raw_length:
            logging.debug("No message length received, exiting")
            sys.exit(0)
        
        message_length = struct.unpack('@I', raw_length)[0]
        logging.debug(f"Message length: {message_length}")
        
        # Read the message itself
        message = sys.stdin.buffer.read(message_length).decode('utf-8')
        logging.debug(f"Received message: {message}")
        
        return json.loads(message)
    except Exception as e:
        logging.error(f"Error reading message: {e}")
        return None

def send_message(message):
    """Send a message to stdout using Chrome native messaging protocol"""
    try:
        encoded = json.dumps(message).encode('utf-8')
        
        # Write message length
        sys.stdout.buffer.write(struct.pack('@I', len(encoded)))
        
        # Write the message
        sys.stdout.buffer.write(encoded)
        sys.stdout.buffer.flush()
        
        logging.debug(f"Sent message: {message}")
    except Exception as e:
        logging.error(f"Error sending message: {e}")

def get_vscode_port():
    """Read the VS Code extension port from the port file"""
    port_file = Path.home() / '.bergamot' / 'port.json'
    
    try:
        if port_file.exists():
            with open(port_file, 'r') as f:
                data = json.load(f)
                return data.get('port', 5000)
    except Exception as e:
        logging.error(f"Error reading port file: {e}")
    
    # Default to port 5000 if file doesn't exist
    return 5000

def forward_to_vscode(message):
    """Forward a message to the VS Code extension HTTP server"""
    port = get_vscode_port()
    
    try:
        # Extract the endpoint and data from the message
        endpoint = message.get('endpoint', '/visit')
        data = message.get('data', {})
        
        # Make HTTP request to VS Code extension
        url = f'http://localhost:{port}{endpoint}'
        logging.info(f"Forwarding to VS Code: {url}")
        
        response = requests.post(
            url,
            json=data,
            headers={'Content-Type': 'application/json'},
            timeout=5
        )
        
        if response.ok:
            logging.info(f"Successfully forwarded to VS Code")
            return {
                'success': True,
                'status': response.status_code,
                'data': response.json() if response.text else None
            }
        else:
            logging.error(f"VS Code returned error: {response.status_code}")
            return {
                'success': False,
                'error': f'VS Code returned status {response.status_code}'
            }
            
    except requests.exceptions.ConnectionError:
        logging.error("Cannot connect to VS Code extension")
        return {
            'success': False,
            'error': 'Cannot connect to VS Code extension. Is it running?'
        }
    except requests.exceptions.Timeout:
        logging.error("Request to VS Code timed out")
        return {
            'success': False,
            'error': 'Request to VS Code extension timed out'
        }
    except Exception as e:
        logging.error(f"Error forwarding to VS Code: {e}")
        return {
            'success': False,
            'error': str(e)
        }

def main():
    """Main message loop"""
    logging.info("Native host started")
    
    while True:
        message = read_message()
        
        if not message:
            logging.error("Invalid message received")
            continue
        
        # Handle different message types
        msg_type = message.get('type')
        
        if msg_type == 'ping':
            # Simple ping/pong for testing
            send_message({'type': 'pong', 'echo': message.get('data')})
            
        elif msg_type == 'get_port':
            # Return the current VS Code port
            port = get_vscode_port()
            send_message({'type': 'port', 'port': port})
            
        elif msg_type == 'forward':
            # Forward message to VS Code extension
            result = forward_to_vscode(message)
            send_message({'type': 'forward_result', **result})
            
        elif msg_type == 'check_status':
            # Check if VS Code extension is running
            port = get_vscode_port()
            try:
                response = requests.get(f'http://localhost:{port}/status', timeout=2)
                is_running = response.ok
            except:
                is_running = False
            
            send_message({
                'type': 'status',
                'vscode_running': is_running,
                'port': port
            })
        else:
            logging.warning(f"Unknown message type: {msg_type}")
            send_message({
                'type': 'error',
                'error': f'Unknown message type: {msg_type}'
            })

if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        logging.error(f"Fatal error in native host: {e}")
        sys.exit(1)